src/wallet/wallet.cpp

```diff
@@ -264,15 +292,15 @@ bool CWallet::AddKeyPubKeyWithDB(WalletBatch &batch, const CKey& secret, const C
     // CCryptoKeyStore has no concept of wallet databases, but calls AddCryptedKey
     // which is overridden below.  To avoid flushes, the database handle is
     // tunneled through to it.
-    bool needsDB = !encrypted_batch;
-    if (needsDB) {
-        encrypted_batch = &batch;
-    }
-    if (!CCryptoKeyStore::AddKeyPubKey(secret, pubkey)) {
-        if (needsDB) encrypted_batch = nullptr;
-        return false;
+    {
+        std::unique_ptr<WithScopedValue<WalletBatch*>> set_encrypted_batch;
```

Comment:
Why is `set_encrypted_batch` created in the case of `encrypted_batch != nullptr`?

Reply:
I don't see a way to get the required construction/destruction points with RAII without having this variable (a local variable just inside the conditional block won't work as it will get destructed too soon).  Note that the code as it is creates an empty `std::unique_ptr` unconditionally (here), and only creates the `WithScopedValue` object if `encrypted_batch == nullptr`.  That gives us the desired RAII semantics.

Reply:
Thanks for clarifying!